File: rust/rle_encode.rs
use std::io::{Read, Write};

fn main() {
    let mut input = Vec::new();
    std::io::stdin().read_to_end(&mut input).unwrap();

    if input.is_empty() {
        return;
    }

    let mut output = Vec::with_capacity(input.len() * 2);
    let mut cur = input[0];
    let mut cnt: u8 = 1;

    for &x in &input[1..] {
        if x == cur {
            if cnt < 255 {
                cnt += 1;
            } else {
                output.push(cnt);
                output.push(cur);
                cnt = 1;
            }
        } else {
            output.push(cnt);
            output.push(cur);
            cur = x;
            cnt = 1;
        }
    }

    output.push(cnt);
    output.push(cur);

    std::io::stdout().write_all(&output).unwrap();
}

File: go/rle_encode.go
package main

import (
	"io"
	"os"
)

func main() {
	input, err := io.ReadAll(os.Stdin)
	if err != nil {
		os.Exit(1)
	}
	if len(input) == 0 {
		return
	}

	out := make([]byte, 0, len(input)*2)
	cur := input[0]
	var cnt byte = 1

	for i := 1; i < len(input); i++ {
		x := input[i]
		if x == cur {
			if cnt < 255 {
				cnt++
				continue
			}
			out = append(out, cnt, cur)
			cnt = 1
			continue
		}
		out = append(out, cnt, cur)
		cur = x
		cnt = 1
	}

	out = append(out, cnt, cur)

	if _, err := os.Stdout.Write(out); err != nil {
		os.Exit(1)
	}
}
